solutions/usaco-616.mdx

```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
-
+#include<bits/stdc++.h>
 using namespace std;
 
-typedef long long ll;
-typedef long double ld;
-
-#define f first;
-#define s second;
-
-#define len(x) (int)x.size();
-
-int main(){
-	freopen("cbarn.in", "r", stdin);
-	freopen("cbarn.out", "w", stdout);
-	ios_base::sync_with_stdio(0); cin.tie(0);
-	int n;
-	cin >> n;
-	vector<int> a(n);
-	int cows = 0;
-	for (int i = 0; i < n; i++) {
-		cin >> a[i];
-		cows += a[i];
-	}
-	int sol = INT_MAX;
-	for (int start = 0; start < n; start++) {
-		int cur = 0, cur_cows = cows;
-		for (int i = 0; i < n; i++) {
-			cur_cows -= a[(start + i) % n];
-			cur += cur_cows;
-		}
-		sol = min(sol, cur);
-	}
-	cout << sol << '\n';
+int main() {
+  freopen("cbarn.in", "r", stdin);
+  freopen("cbarn.out", "w", stdout);
+  cin.sync_with_stdio(0); cin.tie(0);
+
+  // input
+
+  // have the number of cows and desired positions of each cow
+
+  int n;
+  cin >> n;
+
+  vector<int> cows(n);
+
+  for(int i = 0; i < n; i++) {
+    cin >> cows[i];
+  }
+
+  // simple brute force
+
+  // check each possible barn we can open
+
+  int ret = (int)1e9; // stores the current answer
+  // initially set to an abnormally large number
+
+  for(int open_barn = 0; open_barn < n; open_barn++) {
+    int curdist = 0; // stores the current distance if barn i was the opened barn
+    for(int pos = 0; pos < n; pos++) {
+      // the current barn we are processing is open_barn + pos
+      // however, since it is a circular array 
+      // the position may wrap around
+      // index = open_barn + pos
+      curdist += pos * cows[(open_barn + pos) % n];
+    }
+    // update the answer
+    ret = min(ret, curdist);
+  }
+
+  cout << ret << '\n';
+
+  return 0;
 }
 
 ```
 
 </CPPSection>
+	
+<JavaSection>
+	
+## Implementation	
+	
+```java
+import java.io.*;
+import java.util.*;
+public class cbarn {
+	public static void main(String[] args) throws IOException {
+		// initialize file I/O
+		BufferedReader br = new BufferedReader(new FileReader("cbarn.in"));
+		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("cbarn.out")));
+		// read in N
+		int n = Integer.parseInt(br.readLine());
+		int[] cows = new int[n];
+		// read in how many cows need to be in each room
+		for(int i = 0; i < n; i++) {
+			cows[i] = Integer.parseInt(br.readLine());
+		}
+		// the answer cannot exceed N * N * 100, since there are at most 100N cows, each of which can move at most N
+		int ans = n * n * 100;
+		for(int unlock = 0; unlock < n; unlock++) {
+			// assume we unlock the door at index "unlock", compute the distance all cows travel
+			int currentDistance = 0;
+			for(int offset = 0; offset < n; offset++) {
+				// count how many cows have to walk a distance of "offset"
+				currentDistance += offset * cows[(unlock + offset) % n];
+			}
+			// update the answer
+			if(currentDistance < ans) {
+				ans = currentDistance;
+			}
+		}
+		// print the answer
+		pw.println(ans);
+		// close output stream
+		pw.close();
+		br.close();
```

Comment:
lines 78, 81, 94, 99, and 101 are unnecessary comments (they shouldn't explain what you're doing, but explain _why_ you're doing what you're doing!)